crates/sage-cloudpath/src/parquet.rs

```diff
@@ -227,6 +233,130 @@ pub fn serialize_features(
     writer.into_inner()
 }
 
+pub fn build_matched_fragment_schema() -> parquet::errors::Result<Type> {
+    let msg = r#"
+        message schema {
+            required byte_array psm_id (utf8);
+            required byte_array fragment_type (utf8);
+            required int32 fragment_ordinals;
+            required int32 fragment_charge;
+            required float fragment_mz_experimental;
+            required float fragment_mz_calculated;
+            required float fragment_intensity;
+        }
+    "#;
+
+    // required float fragment_mz_experimental;
+    // required float fragment_mz_calculated;
+    parquet::schema::parser::parse_message_type(msg)
+}
+
+pub fn serialize_matched_fragments(
+    features: &[Feature]
+) -> Result<Vec<u8>, parquet::errors::ParquetError> {
+
+    let schema = build_matched_fragment_schema()?;
+
+    let options = WriterProperties::builder()
+        .set_compression(parquet::basic::Compression::ZSTD(ZstdLevel::try_new(3)?))
+        .build();
+
+    let buf = Vec::new();
+
+    let mut writer = SerializedFileWriter::new(buf, schema.into(), options.into())?;
+
+    for features in features.chunks(65536) {
+
+        let mut rg = writer.next_row_group()?;
+
+        if let Some(mut col) = rg.next_column()? {
+            let psm_ids= features.iter()
+                .map(|f| vec![f.psm_id.to_string().as_bytes().into();  f.fragments.fragment_ordinals.len()])
+                .flatten()
+                .collect::<Vec<ByteArray>>();
+
+            col.typed::<ByteArrayType>()
+                .write_batch(&psm_ids, None, None)?;
+            col.close()?;
+        }
+
+        if let Some(mut col) = rg.next_column()? {
+            let fragment_types= features.iter()
+                .map(|f| f.fragments.kinds.clone())
+                .flat_map(|each_list| each_list.into_iter().map(|kind| match kind {
+                    Kind::A => { "a".to_string().as_bytes().into()}
+                    Kind::B => { "b".to_string().as_bytes().into()}
+                    Kind::C => { "c".to_string().as_bytes().into()}
+                    Kind::X => { "x".to_string().as_bytes().into()}
+                    Kind::Y => { "y".to_string().as_bytes().into()}
+                    Kind::Z => { "z".to_string().as_bytes().into()}
+                }))
+                .collect::<Vec<ByteArray>>();
+            col.typed::<ByteArrayType>()
+                .write_batch(&fragment_types, None, None)?;
+            col.close()?;
+        }
+
+        if let Some(mut col) = rg.next_column()? {
+            let fragment_ordinals = features.iter()
+                .map(|f| f.fragments.fragment_ordinals.clone())
```

Comment:
I think this is an unnecessary clone, should be able to do something like (sic, not in front of compiler)
```rs
features.iter().flat_map(|f| f.fragments.fragment_ordinals.iter()).copied().collect()
```

Reply:
Thank you for your suggestion. I have made the necessary updates accordingly.